include/ghex/unstructured/pattern.hpp

```diff
@@ -0,0 +1,346 @@
+                 * the assumption is that each 2D element is a column, with 'levels' vertical elements.
+                 * WARN: a second index, and therefore a more complex iteration space,
+                 * is needed to handle correctly multiple vertical layers*/
+                class iteration_space {
+
+                    private:
+
+                        int m_partition;
+                        std::vector<index_type, gridtools::ghex::allocator::unified_memory_allocator<index_type>> m_local_index;
+                        std::size_t m_levels;
+
+                    public:
+
+                        // ctors
+                        iteration_space() noexcept = default;
+                        iteration_space(const int partition,
+                                const std::vector<index_type, gridtools::ghex::allocator::unified_memory_allocator<index_type>>& local_index,
+                                const std::size_t levels = 1) noexcept :
+                            m_partition{partition},
+                            m_local_index{local_index},
+                            m_levels{levels} {}
+                        iteration_space(const int partition,
+                                std::vector<index_type, gridtools::ghex::allocator::unified_memory_allocator<index_type>>&& local_index,
+                                const std::size_t levels = 1) noexcept :
+                            m_partition{partition},
+                            m_local_index{std::move(local_index)},
+                            m_levels{levels} {}
+                        // less safe but maybe preferable
+                        // template<typename V>
+                        // iteration_space(const int partition, V&& remote_index) noexcept :
+                        //     m_partition{partition},
+                        //     m_remote_index{std::forward<V>(remote_index)} {}
+                        iteration_space(const int partition, const index_type first, const index_type last, const std::size_t levels = 1) noexcept :
+                            m_partition{partition},
+                            m_local_index{},
+                            m_levels{levels} {
+                            m_local_index.resize(static_cast<std::size_t>(last - first + 1));
+                            for (index_type idx = first; idx <= last; ++idx) {
+                                m_local_index[static_cast<std::size_t>(idx)] = idx;
+                            }
+                        }
+
+                        // member functions
+                        int partition() const noexcept { return m_partition; }
+                        // std::vector<index_type, gridtools::allocator::unified_memory_allocator<index_type>>& local_index() noexcept { return m_local_index; }
+                        const std::vector<index_type, gridtools::ghex::allocator::unified_memory_allocator<index_type>>& local_index() const noexcept { return m_local_index; }
```

Comment:
Calling this `local_index` makes me think it's a single value, while the use should be `obj.local_index[i]`. Should we make clear from the name that this returns a container?

Reply:
True. I kept this from Atlas convention, in which all objects like, e,g, partition, remote_index ... are left singular since they are actually `fields`. I can change this everywhere, even if I think I am going to refactor it a little bit on a following PR for a more general approach.